### DISCUSSION TITLE
500 Internal Server Error shows in random moments from random endpoint methods

### DISCUSSION THREAD

**Discussion by nabiedma** [vercel/community#6828] [Help]
### Summary

So, basically we have deployed our finished project to Vercel. Locally the project works great. We are using Next.js 14.1 (but with Pages Router) and using `Lucia Auth` for our Authentication with a `MongoDB` database. In localhost, the project runs great and without errors. But once deployed into Vercel (have double checked to have the correct .env variables), when logging into the project, 1 of 2 endpoints that are being called from the `/api` routes "crashes" for no reason. Taking a look at the logs, they are not very useful since they only show:
`500 Internal Server Error`
`Node.js process exited with exit status: 129. The logs above can help with debugging the issue.`

![image](https://github.com/vercel/community/assets/44773084/8deba079-77f3-4d76-9338-0251f7e606b9)

![image](https://github.com/vercel/community/assets/44773084/cdb2c725-cc9d-4984-9d6a-a924f0aada2c)

And there is no more log info. Last week it was showing more information, couldn't get a copy or screenshot of that but it was also telling something about a response from node that was being called after it was already ended (`response.end()`)

### Steps to Reproduce

The endpoints that give errors (sometimes is one, sometimes is the other one) are coded liked this:

**/api/v1/users/getAllUsersButAdmins**

```
import connectDB from "@/lib/db";
import UserService from "@/db/services/user";
import type { NextApiRequest, NextApiResponse } from "next";

export default async function handler(
  req: NextApiRequest,
  res: NextApiResponse
) {
  if (req.method === "GET") {
    try {
      await connectDB();
      const userService = new UserService();
      const users = await userService.getAllUsersButAdmins();

      if (!users) res.status(404).json({ message: "Users not found" });
      res.status(200).json(users);
    } catch (e) {
      console.error(e);
    }
  }
}

```

**/api/v1/gyms**

```
import connectDB from "@/lib/db";
import type { NextApiRequest, NextApiResponse } from "next";
import GymService from "@/db/services/gym";

export default async function handler(
  req: NextApiRequest,
  res: NextApiResponse
) {
  if (req.method === "GET") {
    try {
      await connectDB();
      const gymService = new GymService();
      const gyms = await gymService.getAllGyms();

      if (!gyms) res.status(404).json({ message: "Gyms not found" });
      res.status(200).json(gyms);
    } catch (e) {
      console.error(e);
    }
  }
}
```
I have already tried changing the final responses for something like:
```
res.status(200).json(gyms);
res.end();
```

But it is the same thing. The crazy thing is that this is also happening in random moments. Maybe I have navigated to other pages of the project, and when I come back to the main page, some endpoint crashes. And locally they work smoothly and correctly.
Maybe it's also useful to say that we are using `Node 20.9` locally to run the project.
Have searched for other similar posts but they do not have answers from Vercel team or they don't get to any solution.

Thanks in advance for the time!

**Comment by QuiiBz** [comment]
Hi, could you try to redeploy your application and let us know if you're still having issues? We've recently fixed an edge-case regarding applications that returned a `connection: keep-alive` header, but I'm not sure if your issue is related.

You mentioned that the applications crashes with an exit code of 129, which means your code raised an [uncaught exception](https://nodejs.org/api/process.html#event-uncaughtexception).

**Reply by nabiedma** [reply]
Hey @QuiiBz, thanks for writing a reply.
Just redeployed it, and tested, but the same keeps happening. Something weird is I do not have any log tracked now 🤔 In other moments I could see instantly the logs of the error and other type of logs. But now I don't see anything.
![image](https://github.com/vercel/community/assets/44773084/f3c3483d-ace1-4ad5-9757-0c9bd46e03ce)

Will check again later. But unfortunately this edge case you have mentioned don't seem to solve this strange issue. Thanks!

**Reply by nabiedma** [reply]
On the logs page, now I am getting this response:

```
Uncaught Exception: Error [ERR_STREAM_WRITE_AFTER_END]: write after end
at ServerResponse.end (node:_http_outgoing:1047:15)
at K.r.end (/var/task/node_modules/next/dist/compiled/next-server/pages-api.runtime.prod.js:20:13470)
at /var/task/node_modules/next/dist/compiled/next-server/pages-api.runtime.prod.js:20:14284
at K.r.send (/var/task/node_modules/next/dist/compiled/next-server/pages-api.runtime.prod.js:20:14293)
at K.r.json (/var/task/node_modules/next/dist/compiled/next-server/pages-api.runtime.prod.js:20:14376)
at l (/var/task/.next/server/pages/api/v1/auth/login.js:1:5237)
at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
at async K (/var/task/node_modules/next/dist/compiled/next-server/pages-api.runtime.prod.js:20:16545)
at async U.render (/var/task/node_modules/next/dist/compiled/next-server/pages-api.runtime.prod.js:20:16981)
at async r3.runApi (/var/task/node_modules/next/dist/compiled/next-server/server.runtime.prod.js:17:41752) {
code: 'ERR_STREAM_WRITE_AFTER_END'
}
Node.js process exited with exit status: 129. The logs above can help with debugging the issue.
```

If it's useful, it's happening on the:
`/api/v1/gyms` `GET` endpoint

**Reply by nabiedma** [reply]
Still waiting for a reply...

We tried changing all the Axios request to the API from axios to fetch(), but it's still the same. We keep receiving 500 errors from random endpoints, sometimes at the beginning of the application or anywhere.
Is there a way we can have a better log of this? This is really frustrating.

Thanks

**Reply by luthfirobbaniy** [reply]
Hello @nabiedma, I also encountered a this similar "random" error and have solved it (even though it's not Next.js). This is the solution that worked for me:

Is your MongoDB server located far away from your deployed app in Vercel? If this true, You can try to move them as close as possible so that your server and app have closer locations (to reduce latency). You can refer to this [link](https://vercel.com/docs/functions/configuring-functions/region#setting-your-default-region) about how to change the region from the Vercel dashboard. Also [add](https://vercel.com/docs/functions/configuring-functions/region#project-configuration) "regions" to your vercel.json configuration file.

```
// vercel.json
{
  "regions": ["sfo1"] // use your preferred region code
}
```

**Reply by QuiiBz** [reply]
Could you share a minimal reproduction? Looking at the error (`ERR_STREAM_WRITE_AFTER_END`), it seems to be caused by something in your code, more specifically at `api/v1/auth/login.js:1:5237`.

**Reply by bikramggg** [reply]
I am also facing the same issue from yesterday. Can someone help please?

[a comment by an automated account is not shown]
